pkg/trillianclient/trillian_client.go

```diff
@@ -75,18 +185,170 @@ func unmarshalLogRoot(logRoot []byte) (types.LogRootV1, error) {
 	return root, nil
 }
 
-func (t *TrillianClient) root(ctx context.Context) (types.LogRootV1, error) {
-	rqst := &trillian.GetLatestSignedLogRootRequest{
-		LogId: t.logID,
-	}
-	resp, err := t.client.GetLatestSignedLogRoot(ctx, rqst)
+// ensureStarted initializes the shared LogClient and starts the updater once.
+//
+// Locking strategy: this uses a double-checked initialization gate so network
+// calls do not hold the mutex. Only one goroutine performs the initial RPCs,
+// others wait on the condition variable until initialization completes. This
+// avoids head-of-line blocking while keeping state updates atomic.
+func (t *TrillianClient) ensureStarted(ctx context.Context) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.started {
+		return t.startErr
+	}
+
+	// Perform one-time initialization while holding the lock for simplicity.
+	// This blocks other ensureStarted callers until initialization completes.
+	cctx := ctx
+	var cancel context.CancelFunc
+	if _, ok := ctx.Deadline(); !ok {
+		cctx, cancel = context.WithTimeout(ctx, t.config.InitLatestRootTimeout)
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+	slr, err := t.client.GetLatestSignedLogRoot(cctx, &trillian.GetLatestSignedLogRootRequest{LogId: t.logID})
 	if err != nil {
-		return types.LogRootV1{}, err
+		t.startErr = err
+		return err
+	}
+	if slr == nil || slr.SignedLogRoot == nil {
+		err = fmt.Errorf("nil signed log root")
+		t.startErr = err
+		return err
+	}
+	r, uerr := unmarshalLogRoot(slr.SignedLogRoot.LogRoot)
+	if uerr != nil {
+		t.startErr = uerr
+		return uerr
+	}
+
+	t.v = client.NewLogVerifier(rfc6962.DefaultHasher)
+	t.lc = client.New(t.logID, t.client, t.v, r)
+	t.snapshot.Store(rootSnapshot{root: r, signed: slr.SignedLogRoot})
+	t.started = true
+	t.startErr = nil
+
+	// Start updater and track it for graceful shutdown.
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+		t.updater()
+	}()
+	return nil
+}
+
+// updater waits for root changes using the LogClient and notifies waiters.
+// It uses the parsed root from WaitForRootUpdate and synthesizes a minimal
+// SignedLogRoot (LogRoot bytes only) to avoid an extra network round trip
+// per advancement.
+func (t *TrillianClient) updater() {
+	// Create backoff for retry logic with reasonable defaults
+	bo := backoff.Backoff{
+		Min:    100 * time.Millisecond, // Start with 100ms
+		Max:    30 * time.Second,       // Cap at 30s
+		Factor: 2.0,                    // Double each time
+		Jitter: true,                   // Add randomization
+	}
+	for {
+		// Wrap the WaitForRootUpdate call with backoff retry
+		var nr *types.LogRootV1
+		err := bo.Retry(t.bgCtx, func() error {
+			select {
+			case <-t.stopCh:
+				return fmt.Errorf("client stopped")
```

Comment:
Should this be returning nil rather than an error, since this is expected on server shutdown?

Reply:
the return value here stops the Retry loop from continuing, which is what we'd want in a shutdown scenario.